File: 2025/java/src/test/java/nl/dirkgroot/adventofcode/year2025/Day01Test.java
package nl.dirkgroot.adventofcode.year2025;

import nl.dirkgroot.adventofcode.util.AoCTest;
import org.junit.jupiter.api.Test;

import java.util.stream.Gatherers;

import static org.assertj.core.api.Assertions.assertThat;

public class Day01Test implements AoCTest {
    record State(int dial, int stoppedAtZero, int passedZero) {
    }

    long part1(String input) {
        return doRotations(input).stoppedAtZero;
    }

    long part2(String input) {
        return doRotations(input).passedZero;
    }

    private static State doRotations(String input) {
        return input.lines()
                .map(line -> line.charAt(0) == 'L'
                        ? -Integer.parseInt(line.substring(1))
                        : Integer.parseInt(line.substring(1)))
                .gather(Gatherers.fold(() -> new State(50, 0, 0), (state, rotate) -> {
                    var dial = state.dial;
                    var step = (rotate > 0) ? 1 : -1;
                    var passedZero = 0;

                    for (var i = 0; i < Math.abs(rotate); i++) {
                        dial = (dial + step) % 100;
                        if (dial == 0) passedZero++;
                    }

                    return new State(dial, state.stoppedAtZero + (dial == 0 ? 1 : 0), state.passedZero + passedZero);
                }))
                .findFirst().orElseThrow();
    }

    static final String EXAMPLE = """
            L68
            L30
            R48
            L5
            R60
            L55
            L1
            L99
            R14
            L82""";

    @Test
    public void part1Example() {
        assertThat(invokeWith(this::part1, EXAMPLE)).isEqualTo(3);
    }

    @Test
    public void part1Solution() {
        assertThat(invokeWith(this::part1, input(1))).isEqualTo(1043);
    }

    @Test
    public void part2Example() {
        assertThat(invokeWith(this::part2, EXAMPLE)).isEqualTo(6);
    }

    @Test
    public void part2Solution() {
        assertThat(invokeWith(this::part2, input(1))).isEqualTo(5963);
    }
}

File: 2024/kotlin/src/test/kotlin/Day01Test.kt
import assertk.assertThat
import assertk.assertions.isEqualTo
import util.input
import util.invokedWith
import org.junit.jupiter.api.Test
import kotlin.math.abs

private fun part1(input: String) =
    parse(input).let { (left, right) ->
        left.sorted().zip(right.sorted())
            .sumOf { (id1, id2) -> abs(id1 - id2) }
    }

private fun part2(input: String) =
    parse(input).let { (left, right) ->
        left.sumOf { id1 -> id1 * right.count { id2 -> id2 == id1 } }
    }

private fun parse(input: String) =
    input.lineSequence()
        .map { it.split("   ") }
        .map { it[0].toInt() to it[1].toInt() }
        .fold(listOf<Int>() to listOf<Int>()) { (left, right), (id1, id2) -> left + id1 to right + id2 }

class Day01Test {
    @Test
    fun `part1 solution`() = assertThat(::part1 invokedWith input(1)).isEqualTo(2113135)

    @Test
    fun `part2 solution`() = assertThat(::part2 invokedWith input(1)).isEqualTo(19097157)
}
